### PR TITLE
First cut of geometric asian pricer

### PR DESCRIPTION
**Pull request by jackgillett101**
Geometric asian option pricer under black scholes, this pricer transforms the asian problem into an equivalent vanilla problem and uses the vanilla pricer to price.

Affects this issue: https://github.com/google/tf-quant-finance/issues/56

### CODE REVIEW

File: tf_quant_finance/black_scholes/asian_prices.py [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+      `volatilities`. The expiry of each option. The units should be such that
+      `expiry * volatility**2` is dimensionless.
+    spots: A real `Tensor` of any shape that broadcasts to the shape of the
+      `volatilities`. The current spot price of the underlying. Either this
+      argument or the `forwards` (but not both) must be supplied.
+    forwards: A real `Tensor` of any shape that broadcasts to the shape of
+      `volatilities`. The forwards to maturity. Either this argument or the
+      `spots` must be supplied but both must not be supplied.
+    sampling_times: A real `Tensor` of same dtype as expiries, n'th column is
+      the sampling times of the n'th option
```
Comment (cyrilchim) [inline tf_quant_finance/black_scholes/asian_prices.py:95]: please specify the shape of `sampling_times`

File: tf_quant_finance/black_scholes/asian_prices.py [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+    forwards = np.array([1.0, 2.0, 3.0, 4.0, 5.0])
+    # Strikes will automatically be broadcasted to shape [5].
+    strikes = np.array([3.0])
+    # Expiries will be broadcast to shape [5], i.e. each option has strike=3
+    # and expiry = 1.
+    expiries = 1.0
+    sampling_times = np.array([[0.5, 0.5, 0.5, 0.5, 0.5], [1.0, 1.0, 1.0, 1.0, 1.0]])
+    past_fixings = np.array([[1.0, 2.0, 3.0, 4.0, 5.0]])
+    computed_prices = tff.black_scholes.discrete_geometric_average_price(
+        volatilities=volatilities,
```
Comment (cyrilchim) [inline tf_quant_finance/black_scholes/asian_prices.py:67]: I think `discrete_geometric_average_price` is a bit vague without referencing Asian options. How about simply  `asian_option_price`? You could add enums for aggregation type as well as option type (discrete/continuous)

File: tf_quant_finance/black_scholes/asian_prices.py [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+      running_accumulator = tf.reduce_prod(past_fixings, 0)
+      fixing_count = past_fixings.shape[0]
+
+    sample_count = sampling_times.shape[0] + fixing_count
+    sampling_time_forwards = (spots * tf.exp((discount_rates - dividend_rates)
+                                              * sampling_times))
+
+    t1 = tf.reduce_sum(sampling_times, 0) / sample_count
+    t2 = tf.reduce_sum(tf.map_fn(lambda x: tf.minimum(*tf.meshgrid(x, tf.transpose(x))),
+                                 tf.transpose(sampling_times)), [1,2]) / sample_count**2
```
Comment (cyrilchim) [inline tf_quant_finance/black_scholes/asian_prices.py:200]: here use tf.vectorized_map and set `fallback_to_while_loop` to False. This should automatically vectorize what you want to do and provide significant speed up on a GPU device

Comment (cyrilchim) [reply tf_quant_finance/black_scholes/asian_prices.py:200]: also, could you please add a comment explaining what is achieved here?

File: tf_quant_finance/black_scholes/asian_prices.py [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+
+    asian_forwards = (tf.math.pow((running_accumulator
+                                   * tf.reduce_prod(sampling_time_forwards, axis=0)),
+                                  1/sample_count)
+                      * tf.math.exp(-0.5*volatilities*volatilities*(t1-t2)))
```
Comment (cyrilchim) [inline tf_quant_finance/black_scholes/asian_prices.py:205]: nit: spaces around `-` in `t1 - t2`

File: tf_quant_finance/black_scholes/asian_prices_test.py [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+    """
+    spots = np.array([100.0])
+    dividend_rates = np.array([0.03])
+    discount_rates = np.array([0.06])
+    volatilities = np.array([0.2])
+    strikes = np.array([100.0])
+    expiries = 1.0
+    sampling_times = np.array([np.linspace(0, 1, 11)[1:]]).transpose()
+
+    expected_price = np.array([5.3425606635])
```
Comment (cyrilchim) [inline tf_quant_finance/black_scholes/asian_prices_test.py:45]: could you please add a comment where this expected price is coming from?

File: tf_quant_finance/black_scholes/asian_prices_test.py [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+  def test_option_prices(self):
+    """Tests for methods for the asian pricing module. Results come from
+    'Implementing Derivatives Model' by Clewlow, Strickland p.118-123, via the
+    QuantLib test suite
+    """
+    spots = np.array([100.0])
+    dividend_rates = np.array([0.03])
+    discount_rates = np.array([0.06])
+    volatilities = np.array([0.2])
+    strikes = np.array([100.0])
```
Comment (cyrilchim) [inline tf_quant_finance/black_scholes/asian_prices_test.py:41]: let's also make it work with scalar inputs:  spots = 100, divident_rates = 0.03, ...
you might need to check rank of inputs (e.g., spots.shape.rank) and see if that is zero, add an extra dimension (e.g., spots = spots[tf.newaxis])

File: tf_quant_finance/black_scholes/asian_prices_test.py [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+    expiries = 1.0
+    sampling_times = np.array([np.linspace(0, 1, 11)[1:]]).transpose()
```
Comment (cyrilchim) [inline tf_quant_finance/black_scholes/asian_prices_test.py:43]: np.linspace(0.1, 1, 10)[:, np.newaxis] is more readable?

**Review by cyrilchim — COMMENTED** [review]
Thank you! This is a very nice contributions. Just a bunch of minor comments

File: tf_quant_finance/black_scholes/asian_prices.py [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+                                     past_fixings: types.RealTensor = None,
+                                     discount_rates: types.RealTensor = None,
+                                     dividend_rates: types.RealTensor = None,
+                                     discount_factors: types.RealTensor = None,
+                                     is_call_options: types.BoolTensor = None,
+                                     is_normal_volatility: bool = False,
+                                     dtype: tf.DType = None,
+                                     name: str = None) -> types.RealTensor:
+  """Computes the Black Scholes price for a batch of geometric averaging,
+  discrete time sampled call or put options.
```
[the remark this answers is not in the document]
Comment (jackgillett101) [reply tf_quant_finance/black_scholes/asian_prices.py]: Fixed

File: tf_quant_finance/black_scholes/asian_prices.py (hunk 1 above)
Comment (jackgillett101) [reply tf_quant_finance/black_scholes/asian_prices.py]: Fixed

File: tf_quant_finance/black_scholes/asian_prices.py (hunk 2 above)
Comment (jackgillett101) [reply tf_quant_finance/black_scholes/asian_prices.py]: Great feedback - this is done now. Pricing for more variants coming soon.

File: tf_quant_finance/black_scholes/asian_prices.py (hunk 3 above)
Comment (jackgillett101) [reply tf_quant_finance/black_scholes/asian_prices.py]: Moved to vectorized_map

I've added an explanation comment above - let me know if this works or if you have a standard way of formatting math comments.

File: tf_quant_finance/black_scholes/asian_prices_test.py (hunk 6 above)
Comment (jackgillett101) [reply tf_quant_finance/black_scholes/asian_prices_test.py:40]: Added a test to do this

File: tf_quant_finance/black_scholes/asian_prices_test.py (hunk 7 above)
Comment (jackgillett101) [reply tf_quant_finance/black_scholes/asian_prices_test.py]: Fixed

File: tf_quant_finance/black_scholes/asian_prices_test.py (hunk 5 above)
Comment (jackgillett101) [reply tf_quant_finance/black_scholes/asian_prices_test.py:44]: There is a comment in line 33-35 of this test that has the source of this price. Unfortunately I don't have access to the textbook it references, I've actually taken this test from QuantLib, here: https://github.com/lballabio/QuantLib/blob/7f69bb6fbe4ac76f0262b717cbf97c6dcd7b69d9/test-suite/asianoptions.cpp#L374

Most of the papers that do discuss geometric options unfortunately don't have sample prices for them. Instead, I've added a new test that compares the analytic pricing to a monte carlo calculation.

**Review by zainaoo — APPROVED** [review]

File: tf_quant_finance/black_scholes/asian_prices.py [hunk 9]
```diff
@@ -0,0 +1,291 @@
...
+
+    sample_count = sampling_times.shape[0] + fixing_count
+    sampling_time_forwards = (
+      spots * tf.exp((discount_rates - dividend_rates) * sampling_times)
+    )
+
+    # We can price a discrete geometric asian option under BS using a vanilla
+    # pricer, if we re-express the following parameters:
+    #
+    # t1 = {\frac 1 n} \sum_{i=1}^n t_i
```
Comment (cyrilchim) [inline tf_quant_finance/black_scholes/asian_prices.py:251]: nit: \frac{1}{n}. Here and below

**Review by cyrilchim — COMMENTED** [review]
Thank you! Could you please squash the commits so I can push to the library.

File: tf_quant_finance/black_scholes/asian_prices.py (hunk 9 above)
Comment (jackgillett101) [reply tf_quant_finance/black_scholes/asian_prices.py]: Fixed

**Pull request merged**